File: threads/src/main/java/im/threads/internal/chat_updates/ChatUpdateProcessor.java
package im.threads.internal.chat_updates;

import androidx.annotation.NonNull;

import java.util.List;

import im.threads.internal.formatters.ChatItemType;
import im.threads.internal.model.ChatItem;
import im.threads.internal.model.ClientNotificationDisplayType;
import im.threads.internal.model.QuickReply;
import im.threads.internal.model.SpeechMessageUpdate;
import im.threads.internal.model.Survey;
import im.threads.internal.transport.ChatItemProviderData;
import im.threads.internal.transport.TransportException;
import im.threads.internal.transport.models.AttachmentSettings;
import io.reactivex.processors.FlowableProcessor;
import io.reactivex.processors.PublishProcessor;

public class ChatUpdateProcessor {

    private static ChatUpdateProcessor instance;

    private final FlowableProcessor<String> typingProcessor = PublishProcessor.create();
    private final FlowableProcessor<AttachmentSettings> attachmentSettingsProcessor = PublishProcessor.create();
    private final FlowableProcessor<String> outgoingMessageReadProcessor = PublishProcessor.create();
    private final FlowableProcessor<String> incomingMessageReadProcessor = PublishProcessor.create();
    private final FlowableProcessor<ChatItem> newMessageProcessor = PublishProcessor.create();
    private final FlowableProcessor<ChatItemProviderData> messageSendSuccessProcessor = PublishProcessor.create();
    private final FlowableProcessor<String> messageSendErrorProcessor = PublishProcessor.create();
    private final FlowableProcessor<ChatItemType> removeChatItemProcessor = PublishProcessor.create();
    private final FlowableProcessor<Survey> surveySendSuccessProcessor = PublishProcessor.create();
    private final FlowableProcessor<String> deviceAddressChangedProcessor = PublishProcessor.create();
    private final FlowableProcessor<Boolean> userInputEnableProcessor = PublishProcessor.create();
    private final FlowableProcessor<List<QuickReply>> quickRepliesProcessor = PublishProcessor.create();
    private final FlowableProcessor<ClientNotificationDisplayType> clientNotificationDisplayTypeProcessor = PublishProcessor.create();
    private final FlowableProcessor<SpeechMessageUpdate> speechMessageUpdateProcessor = PublishProcessor.create();

    private final FlowableProcessor<TransportException> errorProcessor = PublishProcessor.create();

    public static ChatUpdateProcessor getInstance() {
        if (instance == null) {
            instance = new ChatUpdateProcessor();
        }
        return instance;
    }

    public void postTyping(@NonNull String clientId) {
        typingProcessor.onNext(clientId);
    }

    public void postAttachmentSettings(@NonNull AttachmentSettings attachmentSettings) {
        attachmentSettingsProcessor.onNext(attachmentSettings);
    }

    public void postOutgoingMessageWasRead(@NonNull String messageId) {
        outgoingMessageReadProcessor.onNext(messageId);
    }

    public void postIncomingMessageWasRead(@NonNull String messageId) {
        incomingMessageReadProcessor.onNext(messageId);
    }

    public void postSpeechMessageUpdate(@NonNull SpeechMessageUpdate speechMessageUpdate) {
        speechMessageUpdateProcessor.onNext(speechMessageUpdate);
    }

    public void postNewMessage(@NonNull ChatItem chatItem) {
        newMessageProcessor.onNext(chatItem);
    }

    public void postChatItemSendSuccess(@NonNull ChatItemProviderData chatItemProviderData) {
        messageSendSuccessProcessor.onNext(chatItemProviderData);
    }

    public void postChatItemSendError(@NonNull String uuid) {
        messageSendErrorProcessor.onNext(uuid);
    }

    public void postRemoveChatItem(@NonNull ChatItemType chatItemType) {
        removeChatItemProcessor.onNext(chatItemType);
    }

    public void postSurveySendSuccess(Survey survey) {
        surveySendSuccessProcessor.onNext(survey);
    }

    public void postDeviceAddressChanged(String deviceAddress) {
        deviceAddressChangedProcessor.onNext(deviceAddress);
    }

    public void postUserInputEnableChanged(Boolean enable) {
        userInputEnableProcessor.onNext(enable);
    }

    public void postQuickRepliesChanged(List<QuickReply> quickReplies) {
        quickRepliesProcessor.onNext(quickReplies);
    }

    public void postClientNotificationDisplayType(ClientNotificationDisplayType type) {
        clientNotificationDisplayTypeProcessor.onNext(type);
    }

    public void postError(@NonNull TransportException error) {
        errorProcessor.onNext(error);
    }

    public FlowableProcessor<String> getTypingProcessor() {
        return typingProcessor;
    }

    public FlowableProcessor<AttachmentSettings> getAttachmentSettingsProcessor() {
        return attachmentSettingsProcessor;
    }

    public FlowableProcessor<String> getOutgoingMessageReadProcessor() {
        return outgoingMessageReadProcessor;
    }

    public FlowableProcessor<String> getIncomingMessageReadProcessor() {
        return incomingMessageReadProcessor;
    }

    public FlowableProcessor<SpeechMessageUpdate> getSpeechMessageUpdateProcessor() {
        return speechMessageUpdateProcessor;
    }

    public FlowableProcessor<ChatItem> getNewMessageProcessor() {
        return newMessageProcessor;
    }

    public FlowableProcessor<ChatItemProviderData> getMessageSendSuccessProcessor() {
        return messageSendSuccessProcessor;
    }

    public FlowableProcessor<String> getMessageSendErrorProcessor() {
        return messageSendErrorProcessor;
    }

    public FlowableProcessor<ChatItemType> getRemoveChatItemProcessor() {
        return removeChatItemProcessor;
    }

    public FlowableProcessor<Survey> getSurveySendSuccessProcessor() {
        return surveySendSuccessProcessor;
    }

    public FlowableProcessor<String> getDeviceAddressChangedProcessor() {
        return deviceAddressChangedProcessor;
    }

    public FlowableProcessor<Boolean> getUserInputEnableProcessor() {
        return userInputEnableProcessor;
    }

    public FlowableProcessor<List<QuickReply>> getQuickRepliesProcessor() {
        return quickRepliesProcessor;
    }

    public FlowableProcessor<ClientNotificationDisplayType> getClientNotificationDisplayTypeProcessor() {
        return clientNotificationDisplayTypeProcessor;
    }

    public FlowableProcessor<TransportException> getErrorProcessor() {
        return errorProcessor;
    }
}

File: threads/src/main/java/im/threads/business/chatUpdates/ChatUpdateProcessor.kt
package im.threads.business.chatUpdates

import im.threads.business.formatters.ChatItemType
import im.threads.business.models.CampaignMessage
import im.threads.business.models.ChatItem
import im.threads.business.models.ChatItemSendErrorModel
import im.threads.business.models.ClientNotificationDisplayType
import im.threads.business.models.FileDescription
import im.threads.business.models.InputFieldEnableModel
import im.threads.business.models.QuickReplyItem
import im.threads.business.models.SpeechMessageUpdate
import im.threads.business.models.Survey
import im.threads.business.transport.ChatItemProviderData
import im.threads.business.transport.TransportException
import im.threads.business.transport.models.Attachment
import im.threads.business.transport.models.AttachmentSettings
import im.threads.business.transport.threadsGate.responses.Status
import io.reactivex.processors.FlowableProcessor
import io.reactivex.processors.PublishProcessor
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.launch

class ChatUpdateProcessor {

    val typingProcessor: FlowableProcessor<String> = PublishProcessor.create()
    val attachmentSettingsProcessor: FlowableProcessor<AttachmentSettings> = PublishProcessor.create()
    val outgoingMessageStatusChangedProcessor: FlowableProcessor<List<Status>> = PublishProcessor.create()
    val incomingMessageReadProcessor: FlowableProcessor<String> = PublishProcessor.create()
    val newMessageProcessor: FlowableProcessor<ChatItem> = PublishProcessor.create()
    val updateAttachmentsProcessor: FlowableProcessor<List<Attachment>> = PublishProcessor.create()
    val messageSendSuccessProcessor: FlowableProcessor<ChatItemProviderData> = PublishProcessor.create()
    val campaignMessageReplySuccessProcessor: FlowableProcessor<CampaignMessage> = PublishProcessor.create()
    val messageSendErrorProcessor: FlowableProcessor<ChatItemSendErrorModel> = PublishProcessor.create()
    val removeChatItemProcessor: FlowableProcessor<ChatItemType> = PublishProcessor.create()
    val surveySendSuccessProcessor: FlowableProcessor<Survey> = PublishProcessor.create()
    val deviceAddressChangedProcessor: FlowableProcessor<String> = PublishProcessor.create()
    val userInputEnableProcessor: FlowableProcessor<InputFieldEnableModel> = PublishProcessor.create()
    val quickRepliesProcessor: FlowableProcessor<QuickReplyItem> = PublishProcessor.create()
    val clientNotificationDisplayTypeProcessor: FlowableProcessor<ClientNotificationDisplayType> = PublishProcessor.create()
    val speechMessageUpdateProcessor: FlowableProcessor<SpeechMessageUpdate> = PublishProcessor.create()
    val attachAudioFilesProcessor: FlowableProcessor<Boolean> = PublishProcessor.create()
    val errorProcessor: FlowableProcessor<TransportException> = PublishProcessor.create()
    val uploadResultProcessor: FlowableProcessor<FileDescription> = PublishProcessor.create()
    val grantedPermissionsFlow: MutableSharedFlow<Int> = MutableSharedFlow()

    /**
     * Передаёт ответы от коллбэков сокетного соединения.
     * Предоставляется с целью обработки ошибок разрыва соединений.
     */
    val socketResponseMapProcessor: FlowableProcessor<Map<String, Any>> = PublishProcessor.create()
    fun postTyping() {
        typingProcessor.onNext("")
    }

    fun postAttachmentSettings(attachmentSettings: AttachmentSettings) {
        attachmentSettingsProcessor.onNext(attachmentSettings)
    }

    fun postOutgoingMessageStatusChanged(statuses: List<Status>) {
        outgoingMessageStatusChangedProcessor.onNext(statuses)
    }

    fun postIncomingMessageWasRead(messageId: String) {
        incomingMessageReadProcessor.onNext(messageId)
    }

    fun postSpeechMessageUpdate(speechMessageUpdate: SpeechMessageUpdate) {
        speechMessageUpdateProcessor.onNext(speechMessageUpdate)
    }

    fun postNewMessage(chatItem: ChatItem) {
        newMessageProcessor.onNext(chatItem)
    }

    fun updateAttachments(attachments: List<Attachment>) {
        updateAttachmentsProcessor.onNext(attachments)
    }

    fun postChatItemSendSuccess(chatItemProviderData: ChatItemProviderData) {
        messageSendSuccessProcessor.onNext(chatItemProviderData)
    }

    fun postChatItemSendError(sendErrorModel: ChatItemSendErrorModel) {
        messageSendErrorProcessor.onNext(sendErrorModel)
    }

    fun postRemoveChatItem(chatItemType: ChatItemType) {
        removeChatItemProcessor.onNext(chatItemType)
    }

    fun postSurveySendSuccess(survey: Survey) {
        surveySendSuccessProcessor.onNext(survey)
    }

    fun postCampaignMessageReplySuccess(campaignMessage: CampaignMessage) {
        campaignMessageReplySuccessProcessor.onNext(campaignMessage)
    }

    fun postDeviceAddressChanged(deviceAddress: String) {
        deviceAddressChangedProcessor.onNext(deviceAddress)
    }

    fun postUserInputEnableChanged(enable: InputFieldEnableModel) {
        userInputEnableProcessor.onNext(enable)
    }

    fun postQuickRepliesChanged(quickReplies: QuickReplyItem) {
        quickRepliesProcessor.onNext(quickReplies)
    }

    fun postClientNotificationDisplayType(type: ClientNotificationDisplayType) {
        clientNotificationDisplayTypeProcessor.onNext(type)
    }

    fun postAttachAudioFile(attached: Boolean) {
        attachAudioFilesProcessor.onNext(attached)
    }

    fun postError(error: TransportException) {
        errorProcessor.onNext(error)
    }

    fun postSocketResponseMap(socketResponseMap: Map<String, Any>) {
        socketResponseMapProcessor.onNext(socketResponseMap)
    }

    fun postUploadResult(uploadResult: FileDescription?) {
        uploadResultProcessor.onNext(uploadResult)
    }

    fun postGrantedPermissions(permissionsRequestCode: Int) {
        CoroutineScope(Dispatchers.Unconfined).launch {
            grantedPermissionsFlow.emit(permissionsRequestCode)
        }
    }
}
